pkg/controller/bastion/options.go

```diff
@@ -0,0 +1,200 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"net"
+
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/extensions"
+)
+
+//Maximum length for "base" name due to fact that we use this name to name other GCP resources,
+//and it's required to fit 63 character length https://cloud.google.com/compute/docs/naming-resources
+const maxLengthForBaseName = 33
+const maxLengthForResource = 63
+
+// Options contains provider-related information required for setting up
+// a bastion instance. This struct combines precomputed values like the
+// bastion instance name with the IDs of pre-existing cloud provider
+// resources, like the Firewall name, subnet name etc.
+type Options struct {
+	Shoot               *gardencorev1beta1.Shoot
+	BastionInstanceName string
+	CIDRs               []string
+	DiskName            string
+	Zone                string
+	Subnetwork          string
+	ProjectID           string
+	Network             string
+	WorkersCIDR         string
+}
+
+type providerStatusRaw struct {
+	Zone string `json:"zone"`
+}
+
+// DetermineOptions determines the required information that are required to reconcile a Bastion on GCP. This
+// function does not create any IaaS resources.
+func DetermineOptions(bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster, projectID string) (*Options, error) {
+	providerStatus, err := getProviderStatus(bastion)
+	if err != nil {
+		return nil, err
+	}
+
+	cidrs, err := ingressPermissions(bastion)
+	if err != nil {
+		return nil, err
+	}
+
+	//Each resource name up to a maximum of 63 characters in GCP
+	//https://cloud.google.com/compute/docs/naming-resources
+	clusterName := cluster.ObjectMeta.Name
+	baseResourceName, err := generateBastionBaseResourceName(clusterName, bastion.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	workersCidr, err := getWorkersCIDR(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	region := cluster.Shoot.Spec.Region
+	return &Options{
+		Shoot:               cluster.Shoot,
+		BastionInstanceName: baseResourceName,
+		Zone:                getZone(cluster, region, providerStatus),
+		DiskName:            DiskResourceName(baseResourceName),
+		CIDRs:               cidrs,
+		Subnetwork:          fmt.Sprintf("regions/%s/subnetworks/%s", region, NodesResourceName(clusterName)),
+		ProjectID:           projectID,
+		Network:             fmt.Sprintf("projects/%s/global/networks/%s", projectID, clusterName),
```

Comment:
What about the case when existing network is specified?

```yaml
apiVersion: gcp.provider.extensions.gardener.cloud/v1alpha1
kind: InfrastructureConfig
networks:
  vpc:
    name: my-vpc
  workers: 10.250.0.0/16
```

In this case we cannot make assumption that the network name will be the Shoot techincalID.

Reply:
refactor in b296c64c